Add integration for Vallox Ventilation Units

andre-richter opened the pull request:
## Description:

This PR adds initial support for [Vallox ventilation units](https://www.vallox.com/en/products/vallox_ventilation_units.html) that are compatible to the [vallox_websocket_api](https://github.com/yozik04/vallox_websocket_api) PyPi library.

The fan component code reuses parts from the Xiaomi miio fan code, which has been used as a template for this integration.

**Related issue (if applicable):** N/A

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#9668

## Example entry for `configuration.yaml` (if applicable):
```yaml
vallox:
  host: IP_ADDRESS
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

andre-richter committed:
Add integration for Vallox Ventilation Units.

andre-richter commented:
@yozik04 FYI.

yozik04 commented:
Ohh. Nice!

yozik04 commented:
I need to make my lib async as well...

andre-richter commented on homeassistant/components/vallox/fan.py:
```diff
@@ -0,0 +1,325 @@
+SPEED_BOOST = 'Boost'
+SPEED_FIREPLACE = 'Fireplace'
+
+SPEED_LIST = [SPEED_HOME, SPEED_AWAY, SPEED_BOOST, SPEED_FIREPLACE]
```
I did it because the HA functions consuming it later have "speed" in their name. So one way or the other we'll have this indirection.

MartinHjelmare replied:
Only speeds in the base fan component are allowed for the set speed service/method and speed property.

https://github.com/home-assistant/home-assistant/blob/22d9bee41aa71e53a93527faa6fe748f32a9b70b/homeassistant/components/fan/__init__.py#L37-L40

See this architecture issue for a future possible change:
https://github.com/home-assistant/architecture/issues/27

andre-richter replied:
If it is a hard requirement to only expose these speed methods, then I fear this is a blocker for the Vallox ventilation unit integration as I picture it. The units are always-on units that constantly supply a whole building with fresh air, and the control interfaces expose four inherent, manufacturer-provided profiles between which the user can switch. Here's a screenshot of OEM web client:

![image](https://user-images.githubusercontent.com/4130005/59962893-bf58bd00-94eb-11e9-90ea-e7a6c59fe743.png)

Controlling your Vallox units from within HA only makes sense if the profile semantics can be kept. Can we make an exception until your linked PR lands?

MartinHjelmare replied:
It's a hard requirement. We are free to map the device modes to the home assistant speed modes, but that might not make sense in all cases.

andre-richter replied:
While I find it a bit stubborn to not allow this temporarily, I will transform profile selection to a service then.
Give me some days to complete this.

Richiems69 commented:
Unsubscribe

On Fri, Jun 21, 2019, 7:25 AM Andre Richter <notifications@github.com>
wrote:

> *@andre-richter* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/vallox/fan.py
> <https://github.com/home-assistant/home-assistant/pull/24660#discussion_r296197624>
> :
>
> > +    DOMAIN as FAN_DOMAIN, SUPPORT_SET_SPEED, FanEntity)
> +from homeassistant.const import ATTR_ENTITY_ID
> +import homeassistant.helpers.config_validation as cv
> +
> +from . import DOMAIN as VALLOX_DOMAIN, MODE_METRIC_KEY
> +
> +_LOGGER = logging.getLogger(__name__)
> +
> +DATA_KEY = 'fan.vallox'
> +
> +SPEED_HOME = 'At Home'
> +SPEED_AWAY = 'Away'
> +SPEED_BOOST = 'Boost'
> +SPEED_FIREPLACE = 'Fireplace'
> +
> +SPEED_LIST = [SPEED_HOME, SPEED_AWAY, SPEED_BOOST, SPEED_FIREPLACE]
>
> I did it because the HA functions consuming it later have "speed" in their
> name. So one way or the other we'll have this indirection.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/24660?email_source=notifications&email_token=AMNH2HZWNYFXWR3EBMURAW3P3S3DNA5CNFSM4HZZT3X2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4IQFTA#discussion_r296197624>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMNH2H73PMMWOUOFJLAYVFLP3S3DNANCNFSM4HZZT3XQ>
> .
>

MartinHjelmare commented on homeassistant/components/fan/services.yaml:
```diff
@@ -211,3 +211,34 @@ wemo_reset_filter_life:
       description: Names of the WeMo humidifier entities (1 or more entity_ids are required).
       example: 'fan.wemo_humidifier'
+
+vallox_set_profile_fan_speed_home:
```
Integration specific services should be added to the integration package and registered under the integration domain.

andre-richter replied:
Will change.
Just for my own understanding: I tried to start from an existing integration, in this case xiaomi miio and wemo, where they did it that way. Was I wrong to assume this is the same case, or did these rules not apply back when they integrated it?

MartinHjelmare replied:
Those integrations are not the best examples, since they don't always follow our current standards and architecture.

Our architecture decisions have evolved and there are still some old integrations that don't follow them.

MartinHjelmare commented on homeassistant/components/vallox/sensor.py:
```diff
@@ -0,0 +1,189 @@
+                      "Check for \"vallox:\" in the configuration.")
+        return
+
+    global SENSOR_INSTANCES
```
Don't use global. We can use hass.data to store instances specific to the integration platform.

MartinHjelmare commented on homeassistant/components/vallox/__init__.py:
```diff
@@ -0,0 +1,80 @@
+        hass.data[DATA_VALLOX_SENSOR_LIST] = sensor_list
+        await discovery.async_load_platform(hass, 'sensor', DOMAIN, {}, config)
+
+    await discovery.async_load_platform(hass, 'fan', DOMAIN, {}, config)
```
Use `hass.async_create_task` to schedule this as a task to avoid possible deadlock.

MartinHjelmare commented on homeassistant/components/vallox/__init__.py:
```diff
@@ -0,0 +1,80 @@
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): vol.All(ipaddress.ip_address, cv.string),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_SENSORS): vol.All(cv.ensure_list, [vol.In(SENSORS)]),
```
We no longer want to include config options for selecting sensor types. All available types should be included by default.

See https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md.

MartinHjelmare commented on homeassistant/components/vallox/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "requirements": [
+    "vallox-websocket-api==1.5.1",
+    "numpy==1.16.3",
+    "websocket-client==0.54.0"
```
Why do we need to include the websocket client as a requirement? I would think that the vallox websocket api library would depend on the websocket client library.

MartinHjelmare replied:
Why include numpy? We don't seem to use that.

yozik04 replied:
vallox-websockets-api lib uses that. But it should be pulled as lib dependency then.

MartinHjelmare reviewed (changes requested):
I've not done a full review yet. I've only commented on the main architectural dependent changes needed.

yozik04 commented:
I am in progress or changing vallox-websocket-api to be async. Maybe will finish next week.

andre-richter commented:
I want to refactor a bit and introduce an intermediary class that fetches and caches all read values with a single call periodically. Fan and sensor classes would then fetch those cached values.
This would reduce the current wasteful number of websocket calls.

@martinhjelmare, which is an integration I can use as reference? Is smartthings okay?

MartinHjelmare commented:
Yes, that's a good example. :+1:

andre-richter commented:
Addressing the review comments resulted in a bigger refactoring. Please have a look. Might still need some tweaks here and there, but I hope I addressed all the bigger architectural concerns.

Removing `numpy` and `websocket-client` from the manifest did not work in my tests. They were not pulled in automatically when executing `hass` for the first time. Instead, an error message stated that loading the `vallox` integration did not work because those packets are missing.
Please advise on that part.

I will update the corresponding documentation PR once we agreed on the architecture.

MartinHjelmare commented on homeassistant/components/vallox/__init__.py:
```diff
@@ -0,0 +1,252 @@
+        _LOGGER.debug("Updating Vallox state cache.")
+
+        try:
+            self._metric_cache = self._client.fetch_metrics()
```
If I'm reading the vallox library correctly, this does non asyncio I/O in a coroutine. That's not allowed. Coroutines will be executed in the event loop which should never block.

MartinHjelmare replied:
Consider switching affected methods to regular sync functions. This might mean we have to use the home assistant sync api.

The alternative is to schedule I/O on the executor thread pool with `hass.async_add_executor_job`. But if we need to do that a lot, the sync api is cleaner.

MartinHjelmare commented on homeassistant/components/vallox/fan.py:
```diff
@@ -0,0 +1,167 @@
+        device = ValloxFan(hass.data[DOMAIN]['name'],
+                           hass.data[DOMAIN]['client'],
+                           hass.data[DOMAIN]['state_proxy'])
+    except KeyError:
```
Remove this. Instead add a guard clause at the top of the function body that checks if discovery_info is None and return if so.

MartinHjelmare commented on homeassistant/components/vallox/__init__.py:
```diff
@@ -0,0 +1,252 @@
+
+async def async_setup(hass, config):
+    """Set up the client and boot the platforms."""
+    if DOMAIN not in config:
```
This can't happen. There is no network discovery and we're not using config entries here yet.

MartinHjelmare commented on homeassistant/components/vallox/fan.py:
```diff
@@ -0,0 +1,167 @@
+        self._fan_speed_boost = None
+
+        try:
+            self._client.set_settable_address(METRIC_KEY_MODE, int)
```
Please avoid side effects in init method.

MartinHjelmare commented on homeassistant/components/vallox/fan.py:
```diff
@@ -0,0 +1,167 @@
+                    self._available = False
+                    _LOGGER.error("Error turning on: %s", io_err)
+            else:
+                _LOGGER.error("Already on.")
```
If we have the wrong state info we won't be able to change state. If that's a possibility, we try to avoid guards like this.

andre-richter replied:
I put it there to catch a call to the service `fan/turn_on`, where tests showed it executes even if the fan is already on.
This could for example happen if a user-automation script is erroneous.

MartinHjelmare replied:
Would it be bad for the device to call turn on if it's already on?

andre-richter replied:
I don't know. There's no documentation that would give it away, so I did not want to take chances and cause side effects in the device when we can prevent it on the higher level.

MartinHjelmare replied:
I think it's better to put protection close to the device if the device requires protection.

andre-richter replied:
The problem is we cannot find out if the device firmware is fine with mutliple calls or if there would be unforeseen consequences if it would be spammed with those commands due to a user script gone rogue (making up a "dramatic" case here).

In the end it is just an if/else on a boolean to be on the safe side here. IMO, that's an acceptable price to pay for a little more safety?

MartinHjelmare commented on homeassistant/components/vallox/fan.py:
```diff
@@ -0,0 +1,167 @@
+        _LOGGER.debug("Turn on: %s", speed)
+
+        # Case speed == None equals the GUI toggle switch being activated.
+        if speed is None:
```
Invert this and make it a guard clause instead. Ie return if true. Then we can outdent below.

MartinHjelmare commented on homeassistant/components/vallox/fan.py:
```diff
@@ -0,0 +1,167 @@
+            if mode == 0:
+                self._state = True
+            else:
+                self._state = None
```
We don't want to set it to `False` which means off?

MartinHjelmare commented:
I'll look into the requirements.

MartinHjelmare commented:
I think the design is ok. Consider using the home assistant sync api until the library supports asyncio.

andre-richter committed:
Address review comments #1

andre-richter commented:
Due to lack of personal bandwidth, I went for `executor_jobs` instead of of changing APIs. Will make transition to the async version of `vallox_websocket_api` in the near future straight forward.

MartinHjelmare commented on homeassistant/components/vallox/__init__.py:
```diff
@@ -0,0 +1,259 @@
+        _LOGGER.debug("Fetching metric key: %s", metric_key)
+
+        if not self._valid:
+            raise IOError("Device state out of sync.")
```
`IOError` is deprecated. Use `OSError` instead.

MartinHjelmare commented on homeassistant/components/vallox/__init__.py:
```diff
@@ -0,0 +1,259 @@
+        if not self._valid:
+            raise IOError("Device state out of sync.")
+
+        if metric_key not in vlxDevConstants.__dict__.keys():
```
Checking dict key membership doesn't require using `.keys()`.

MartinHjelmare commented on homeassistant/components/vallox/fan.py:
```diff
@@ -0,0 +1,170 @@
+        await hass.async_add_executor_job(
+            client.set_settable_address, METRIC_KEY_MODE, int)
+
+    except AttributeError as attr_err:
```
When can this happen?

andre-richter replied:
If wrong arguments would be given: https://github.com/yozik04/vallox_websocket_api/blob/master/vallox_websocket_api/client.py#L163-L178

Could indeed be removed, since it would trigger during local testing. Shall I?

MartinHjelmare replied:
We're in control of the arguments, right? Yes, then I think we can remove this try...except.

MartinHjelmare commented on homeassistant/components/vallox/sensor.py:
```diff
@@ -0,0 +1,209 @@
+            state_proxy=state_proxy,
+            metric_key='A_CYC_REMAINING_TIME_FOR_FILTER',
+            device_class=None,
+            unit_of_measurement='days',
```
We don't want to store state that is relative time. That generates not needed state updates. It's ok to use an absolute utc iso formatted timestamp and set the device class to timestamp. Then the absolute timestamp will show as a relative time in the frontend.

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

andre-richter committed:
Address review comments #2

MartinHjelmare commented on homeassistant/components/vallox/sensor.py:
```diff
@@ -204,6 +205,29 @@ def __init__(self, name, state_proxy, device_class, unit_of_measurement,
+            # Since only a delta of days is received from the device, fix the
+            # time so the timestamp does not change with every update.
+            now = datetime.utcnow().replace(
+                hour=13, minute=0, second=0, microsecond=0)
```
Very good!

MartinHjelmare reviewed (commented):
I think this looks good! I'll have a look at the requirements now.

MartinHjelmare commented on homeassistant/components/vallox/__init__.py:
```diff
@@ -0,0 +1,259 @@
+                self._client.get_profile)
+            self._valid = True
+
+        except IOError as io_err:
```
We can except OSError instead. IOError is an alias of OSError.

https://docs.python.org/3/library/exceptions.html#IOError

> The following exceptions are kept for compatibility with previous versions; starting from Python 3.3, they are aliases of OSError.

```py
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.3.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: try: 
   ...:     raise IOError 
   ...: except OSError: 
   ...:     print("Caught the IOError") 
   ...:                                                                                                                                                                                                       
Caught the IOError

In [2]:
```

MartinHjelmare commented:
I think I've found the problem with the requirements not installing:
https://github.com/yozik04/vallox_websocket_api/pull/9

andre-richter committed:
Replace IOError with OSError.

yozik04 commented:
PR Merged and new version created: https://pypi.org/project/vallox-websocket-api/
Use: 1.5.2

About async version of the API. For some reason aiohttp.ClientRequest does not connect to Vallox websocket... Wireshark shows exactly the same request. Response from the unit is different... Investigating.

andre-richter committed:
Bump to fixed version of vallox_websocket_api.

andre-richter commented:
Tested and working with `vallox_websocket_api` 1.5.2

MartinHjelmare reviewed (approved):
Great!

yozik04 commented:
I have released async version of the API
[vallox_websocket_api  2.0.0](https://pypi.org/project/vallox-websocket-api/)

andre-richter commented:
Hi @yozik04 ,

Please see the error message below from https://community.home-assistant.io/t/vallox-valloplus-350mv-partial-success-work-in-progress-on-vallox-component/52588/30

Is the response from the Vallox unit bad? Anything we can do about it? I hit this bug as well already, sometimes it helps to try multiple times.

```
Error during setup of component vallox
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/websockets/client.py", line 77, in read_http_response
    status_code, headers = yield from read_response(self.reader)
  File "/usr/local/lib/python3.7/site-packages/websockets/http.py", line 123, in read_response
    status_line = yield from read_line(stream)
  File "/usr/local/lib/python3.7/site-packages/websockets/http.py", line 197, in read_line
    raise ValueError("Line without CRLF")
ValueError: Line without CRLF

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 153, in _async_setup_component
    hass, processed_config)
  File "/usr/src/homeassistant/homeassistant/components/vallox/__init__.py", line 114, in async_setup
    await state_proxy.async_update(None)
  File "/usr/src/homeassistant/homeassistant/components/vallox/__init__.py", line 163, in async_update
    self._metric_cache = await self._client.fetch_metrics()
  File "/usr/local/lib/python3.7/site-packages/vallox_websocket_api/client.py", line 235, in fetch_metrics
    result = await self._websocket_request(command=vlxDevConstants.WS_WEB_UI_COMMAND_READ_TABLES)
  File "/usr/local/lib/python3.7/site-packages/vallox_websocket_api/client.py", line 223, in _websocket_request
    async with websockets.connect("ws://%s/" % self.ip_address) as ws:
  File "/usr/local/lib/python3.7/site-packages/websockets/py35/client.py", line 2, in __aenter__
    return await self
  File "/usr/local/lib/python3.7/site-packages/websockets/py35/client.py", line 19, in __await_impl__
    extra_headers=protocol.extra_headers,
  File "/usr/local/lib/python3.7/site-packages/websockets/client.py", line 260, in handshake
    status_code, response_headers = yield from self.read_http_response()
  File "/usr/local/lib/python3.7/site-packages/websockets/client.py", line 79, in read_http_response
    raise InvalidMessage("Malformed HTTP message") from exc
websockets.exceptions.InvalidMessage: Malformed HTTP message
```

MartinHjelmare commented:
Please don't discuss in merged PRs. Open an issue instead. Thanks!
